File: lib/lazy_frame.ml
open! Core

type t

external scan_parquet : string -> (t, string) result = "rust_lazy_frame_scan_parquet"

let scan_parquet_exn path =
  scan_parquet path |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

external scan_csv : string -> (t, string) result = "rust_lazy_frame_scan_csv"

let scan_csv_exn path =
  scan_csv path |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

external scan_jsonl : string -> (t, string) result = "rust_lazy_frame_scan_jsonl"

let scan_jsonl_exn path =
  scan_jsonl path |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

external explain
  :  t
  -> optimized:bool
  -> (string, string) result
  = "rust_lazy_frame_explain"

let explain ?(optimized = true) t = explain t ~optimized

let explain_exn ?optimized t =
  explain ?optimized t |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

external to_dot
  :  t
  -> optimized:bool
  -> (string, string) result
  = "rust_lazy_frame_to_dot"

let to_dot ?(optimized = true) t = to_dot t ~optimized

let to_dot_exn ?optimized t =
  to_dot ?optimized t |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

external cache : t -> t = "rust_lazy_frame_cache"

external collect
  :  t
  -> streaming:bool
  -> release_runtime:bool
  -> (Data_frame0.t, string) result
  = "rust_lazy_frame_collect"

external collect_all
  :  t list
  -> release_runtime:bool
  -> (Data_frame0.t list, string) result
  = "rust_lazy_frame_collect_all"

external profile
  :  t
  -> release_runtime:bool
  -> (Data_frame0.t * Data_frame0.t, string) result
  = "rust_lazy_frame_profile"

external fetch
  :  t
  -> n_rows:int
  -> release_runtime:bool
  -> (Data_frame0.t, string) result
  = "rust_lazy_frame_fetch"

type profile_result =
  { collected : Data_frame0.t
  ; profile : Data_frame0.t
  }

module Deferred = struct
  open Async

  let collect ?(streaming = false) t =
    In_thread.run (fun () -> collect t ~streaming ~release_runtime:true)
  ;;

  let collect_exn ?streaming t =
    collect ?streaming t >>| Result.map_error ~f:Error.of_string >>| Or_error.ok_exn
  ;;

  let collect_all t = In_thread.run (fun () -> collect_all t ~release_runtime:true)

  let collect_all_exn ts =
    collect_all ts >>| Result.map_error ~f:Error.of_string >>| Or_error.ok_exn
  ;;

  let profile t =
    let%map result = In_thread.run (fun () -> profile t ~release_runtime:true) in
    let%map.Result collected, profile = result in
    { collected; profile }
  ;;

  let profile_exn t =
    profile t >>| Result.map_error ~f:Error.of_string >>| Or_error.ok_exn
  ;;

  let fetch t ~n_rows = In_thread.run (fun () -> fetch t ~n_rows ~release_runtime:true )

  let fetch_exn t ~n_rows =
    fetch t ~n_rows >>| Result.map_error ~f:Error.of_string >>| Or_error.ok_exn
  ;;
end

let collect ?(streaming = false) t = collect t ~streaming ~release_runtime:false

let collect_exn ?streaming t =
  collect ?streaming t |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

let collect_all t = collect_all t ~release_runtime:false

let collect_all t = collect_all t

let collect_all_exn ts =
  collect_all ts |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

let profile t =
  let%map.Result collected, profile = profile t ~release_runtime:false in
  { collected; profile }
;;

type profile_result =
  { collected : Data_frame0.t
  ; profile : Data_frame0.t
  }

let profile t =
  let%map.Result collected, profile = profile t in
  { collected; profile }
;;

let profile_exn t = profile t |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
let fetch t ~n_rows = fetch t ~n_rows ~release_runtime:false

let fetch t ~n_rows = fetch t ~n_rows

let fetch_exn t ~n_rows =
  fetch t ~n_rows |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn
;;

external filter : t -> predicate:Expr.t -> t = "rust_lazy_frame_filter"
external select : t -> exprs:Expr.t list -> t = "rust_lazy_frame_select"
external with_columns : t -> exprs:Expr.t list -> t = "rust_lazy_frame_with_columns"

external groupby
  :  t
  -> is_stable:bool
  -> by:Expr.t list
  -> agg:Expr.t list
  -> t
  = "rust_lazy_frame_groupby"

let groupby ?(is_stable = false) t ~by ~agg = groupby t ~is_stable ~by ~agg

external groupby_dynamic
  :  t
  -> index_column:Expr.t
  -> by:Expr.t list
  -> every:string option
  -> period:string option
  -> offset:string option
  -> truncate:bool option
  -> include_boundaries:bool option
  -> closed_window:[ `Left | `Right | `Both | `None_ ] option
  -> start_by:
       [ `Window_bound
       | `Data_point
       | `Monday
       | `Tuesday
       | `Wednesday
       | `Thursday
       | `Friday
       | `Saturday
       | `Sunday
       ]
       option
  -> check_sorted:bool option
  -> agg:Expr.t list
  -> t
  = "rust_lazy_frame_groupby_dynamic_bytecode" "rust_lazy_frame_groupby_dynamic"

let groupby_dynamic
  ?every
  ?period
  ?offset
  ?truncate
  ?include_boundaries
  ?closed_window
  ?start_by
  ?check_sorted
  t
  ~index_column
  ~by
  ~agg
  =
  (* Following the logic of:
     https://github.com/pola-rs/polars/blob/a8489558008652fe06e182dbdf082e8d9f0159ab/py-polars/polars/lazyframe/frame.py#L2327
  *)
  let offset =
    Option.value
      offset
      ~default:
        (match period with
         | None -> "-" ^ Option.value_exn every
         | Some _ -> "0ns")
  in
  let period = Option.value period ~default:(Option.value_exn every) in
  groupby_dynamic
    t
    ~index_column
    ~by
    ~every
    ~period:(Some period)
    ~offset:(Some offset)
    ~truncate
    ~include_boundaries
    ~closed_window
    ~start_by
    ~check_sorted
    ~agg
;;

external join_
  :  t
  -> other:t
  -> left_on:Expr.t list
  -> right_on:Expr.t list
  -> how:Join_type.t
  -> t
  = "rust_lazy_frame_join"

let join t ~other ~on ~how = join_ t ~other ~left_on:on ~right_on:on ~how
let join' = join_

external vertical_concat
  :  t list
  -> rechunk:bool
  -> parallel:bool
  -> to_supertypes:bool
  -> t
  = "rust_lazy_frame_vertical_concat"

external diagonal_concat
  :  t list
  -> rechunk:bool
  -> parallel:bool
  -> t
  = "rust_lazy_frame_diagonal_concat"

let concat ?(how = `Vertical) ?(rechunk = false) ?(parallel = false) ts =
  match how with
  | `Vertical -> vertical_concat ts ~rechunk ~parallel ~to_supertypes:false
  | `Vertical_relaxed -> vertical_concat ts ~rechunk ~parallel ~to_supertypes:true
  | `Diagonal -> diagonal_concat ts ~rechunk ~parallel
;;

external melt
  :  t
  -> id_vars:string list
  -> value_vars:string list
  -> variable_name:string option
  -> value_name:string option
  -> streamable:bool
  -> t
  = "rust_lazy_frame_melt_bytecode" "rust_lazy_frame_melt"

let melt ?variable_name ?value_name ?(streamable = false) t ~id_vars ~value_vars =
  melt t ~id_vars ~value_vars ~variable_name ~value_name ~streamable
;;

external sort
  :  t
  -> by_column:string
  -> descending:bool option
  -> nulls_last:bool option
  -> multithreaded:bool option
  -> maintain_order:bool option
  -> t
  = "rust_lazy_frame_sort_bytecode" "rust_lazy_frame_sort"

let sort ?descending ?nulls_last t ~by_column =
  sort
    t
    ~by_column
    ~descending
    ~nulls_last (* TODO: make following parameters configurable *)
    ~multithreaded:(Some false)
    ~maintain_order:(Some true)
;;

external limit : t -> n:int -> t = "rust_lazy_frame_limit"
external explode : t -> columns:Expr.t list -> t = "rust_lazy_frame_explode"
external with_streaming : t -> toggle:bool -> t = "rust_lazy_frame_with_streaming"
external schema : t -> (Schema.t, string) result = "rust_lazy_frame_schema"

let schema_exn t = schema t |> Result.map_error ~f:Error.of_string |> Or_error.ok_exn

File: rust/polars-ocaml/src/lazy_frame.rs
// TODO: rust_lazy_frame_profile's return type is triggering this warning, but
// due to how I've implemented the proc macro specifying the allow clippy
// attribute alongside the proc macro invocation doesn't work, which probably
// is fixable; disabling the warning file-wide in the interim.
#![allow(clippy::type_complexity)]
use crate::utils::*;
use ocaml_interop::{DynBox, OCaml, OCamlInt, OCamlList, OCamlRef, ToOCaml};
use polars::prelude::*;
use polars_ocaml_macros::ocaml_interop_export;
use smartstring::{LazyCompact, SmartString};
use std::rc::Rc;
use std::{cell::RefCell, path::Path};

#[ocaml_interop_export]
fn rust_lazy_frame_scan_csv(
    cr: &mut &mut OCamlRuntime,
    path: OCamlRef<String>,
) -> OCaml<Result<DynBox<LazyFrame>, String>> {
    let path: String = path.to_rust(cr);

    LazyCsvReader::new(path)
        .finish()
        .map(Abstract)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
}

// TODO: properly return error type instead of a string
#[ocaml_interop_export]
fn rust_lazy_frame_scan_parquet(
    cr: &mut &mut OCamlRuntime,
    path: OCamlRef<String>,
) -> OCaml<Result<DynBox<LazyFrame>, String>> {
    let path: String = path.to_rust(cr);
    let path: &Path = Path::new(&path);

    LazyFrame::scan_parquet(path, Default::default())
        .map(Abstract)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
}

// TODO: Polars only has a lazy version of JSON Lines reader (and no lazy version of JSON reader),
// which I think is mainly because in the case of JSON it loads a single array. If we have a
// SAX-like parser for JSON, I don't see why we can't have a lazy JSON reader too.
#[ocaml_interop_export]
fn rust_lazy_frame_scan_jsonl(
    cr: &mut &mut OCamlRuntime,
    path: OCamlRef<String>,
) -> OCaml<Result<DynBox<LazyFrame>, String>> {
    let path: String = path.to_rust(cr);
    let path: &Path = Path::new(&path);

    LazyJsonLineReader::new(path)
        .finish()
        .map(Abstract)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
}

#[ocaml_interop_export]
fn rust_lazy_frame_explain(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    optimized: OCamlRef<bool>,
) -> OCaml<Result<String, String>> {
    let Abstract(lazy_frame) = lazy_frame.to_rust(cr);
    let optimized = optimized.to_rust(cr);

    lazy_frame
        .explain(optimized)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
}

#[ocaml_interop_export]
fn rust_lazy_frame_cache(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
) -> OCaml<DynBox<LazyFrame>> {
    dyn_box!(cr, |lazy_frame| lazy_frame.cache())
}

#[ocaml_interop_export]
fn rust_lazy_frame_to_dot(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    optimized: OCamlRef<bool>,
) -> OCaml<Result<String, String>> {
    let Abstract(lazy_frame) = lazy_frame.to_rust(cr);
    let optimized = optimized.to_rust(cr);

    lazy_frame
        .to_dot(optimized)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
}

#[ocaml_interop_export]
fn rust_lazy_frame_collect(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    streaming: OCamlRef<bool>,
) -> OCaml<Result<DynBox<crate::data_frame::PolarsDataFrame>, String>> {
    let streaming = streaming.to_rust(cr);

    dyn_box_result!(cr, |lazy_frame| {
        cr.releasing_runtime(|| {
            lazy_frame
                .with_streaming(streaming)
                .collect()
                .map(|df| Rc::new(RefCell::new(df)))
        })
    })
}

#[ocaml_interop_export]
fn rust_lazy_frame_collect_all(
    cr: &mut &mut OCamlRuntime,
    lazy_frames: OCamlRef<OCamlList<DynBox<LazyFrame>>>,
) -> OCaml<Result<OCamlList<DynBox<crate::data_frame::PolarsDataFrame>>, String>> {
    let lazy_frames = unwrap_abstract_vec(lazy_frames.to_rust(cr));

    cr.releasing_runtime(|| {
        collect_all(lazy_frames)
            .map(|data_frames| {
                data_frames
                    .into_iter()
                    .map(|df| Abstract(Rc::new(RefCell::new(df))))
                    .collect::<Vec<_>>()
            })
            .map_err(|err| err.to_string())
    })
    .to_ocaml(cr)
}

#[ocaml_interop_export]
fn rust_lazy_frame_profile(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
) -> OCaml<
    Result<
        (
            DynBox<crate::data_frame::PolarsDataFrame>,
            DynBox<crate::data_frame::PolarsDataFrame>,
        ),
        String,
    >,
> {
    let Abstract(lazy_frame) = lazy_frame.to_rust(cr);

    cr.releasing_runtime(|| {
        lazy_frame
            .profile()
            .map(|(materialized, profile)| {
                (
                    Abstract(Rc::new(RefCell::new(materialized))),
                    Abstract(Rc::new(RefCell::new(profile))),
                )
            })
            .map_err(|err| err.to_string())
    })
    .to_ocaml(cr)
}

#[ocaml_interop_export(raise_on_err)]
fn rust_lazy_frame_fetch(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    n_rows: OCamlRef<OCamlInt>,
) -> OCaml<Result<DynBox<crate::data_frame::PolarsDataFrame>, String>> {
    let n_rows = n_rows.to_rust::<Coerce<_, i64, usize>>(cr).get()?;

    dyn_box_result!(cr, |lazy_frame| {
        cr.releasing_runtime(|| lazy_frame.fetch(n_rows).map(|df| Rc::new(RefCell::new(df))))
    })
}

#[ocaml_interop_export]
fn rust_lazy_frame_filter(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    expr: OCamlRef<DynBox<Expr>>,
) -> OCaml<DynBox<LazyFrame>> {
    dyn_box!(cr, |lazy_frame, expr| lazy_frame.filter(expr))
}

#[ocaml_interop_export]
fn rust_lazy_frame_select(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    exprs: OCamlRef<OCamlList<DynBox<Expr>>>,
) -> OCaml<DynBox<LazyFrame>> {
    let exprs = unwrap_abstract_vec(exprs.to_rust(cr));

    dyn_box!(cr, |lazy_frame| lazy_frame.select(&exprs))
}

#[ocaml_interop_export]
fn rust_lazy_frame_with_columns(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    exprs: OCamlRef<OCamlList<DynBox<Expr>>>,
) -> OCaml<DynBox<LazyFrame>> {
    let exprs = unwrap_abstract_vec(exprs.to_rust(cr));

    dyn_box!(cr, |lazy_frame| lazy_frame.with_columns(&exprs))
}

#[ocaml_interop_export]
fn rust_lazy_frame_groupby(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    is_stable: OCamlRef<bool>,
    by: OCamlRef<OCamlList<DynBox<Expr>>>,
    agg: OCamlRef<OCamlList<DynBox<Expr>>>,
) -> OCaml<DynBox<LazyFrame>> {
    let is_stable = is_stable.to_rust(cr);
    let by = unwrap_abstract_vec(by.to_rust(cr));
    let agg = unwrap_abstract_vec(agg.to_rust(cr));

    dyn_box!(cr, |lazy_frame| {
        let groupby = if is_stable {
            lazy_frame.groupby_stable(by)
        } else {
            lazy_frame.groupby(by)
        };
        groupby.agg(agg)
    })
}

#[ocaml_interop_export]
fn rust_lazy_frame_groupby_dynamic(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    index_column: OCamlRef<DynBox<Expr>>,
    by: OCamlRef<OCamlList<DynBox<Expr>>>,
    every: OCamlRef<Option<String>>,
    period: OCamlRef<Option<String>>,
    offset: OCamlRef<Option<String>>,
    truncate: OCamlRef<Option<bool>>,
    include_boundaries: OCamlRef<Option<bool>>,
    closed_window: OCamlRef<Option<ClosedWindow>>,
    start_by: OCamlRef<Option<StartBy>>,
    check_sorted: OCamlRef<Option<bool>>,
    agg: OCamlRef<OCamlList<DynBox<Expr>>>,
) -> OCaml<DynBox<LazyFrame>> {
    let every = every
        .to_rust::<Option<String>>(cr)
        .as_deref()
        .map(Duration::parse);
    let period = period
        .to_rust::<Option<String>>(cr)
        .as_deref()
        .map(Duration::parse);
    let offset = offset
        .to_rust::<Option<String>>(cr)
        .as_deref()
        .map(Duration::parse);
    let truncate: Option<bool> = truncate.to_rust(cr);
    let include_boundaries: Option<bool> = include_boundaries.to_rust(cr);
    let closed_window = closed_window
        .to_rust::<Option<PolarsClosedWindow>>(cr)
        .map(|PolarsClosedWindow(closed_window)| closed_window);
    let start_by = start_by
        .to_rust::<Option<PolarsStartBy>>(cr)
        .map(|PolarsStartBy(start_by)| start_by);
    let check_sorted: Option<bool> = check_sorted.to_rust(cr);

    let options: DynamicGroupOptions = Default::default();
    let options = DynamicGroupOptions {
        // index_column is set within LazyFrame::groupby_dynamic()
        index_column: "".into(),
        every: every.unwrap_or(options.every),
        period: period.unwrap_or(options.period),
        offset: offset.unwrap_or(options.offset),
        truncate: truncate.unwrap_or(options.truncate),
        include_boundaries: include_boundaries.unwrap_or(options.include_boundaries),
        closed_window: closed_window.unwrap_or(options.closed_window),
        start_by: start_by.unwrap_or(options.start_by),
        check_sorted: check_sorted.unwrap_or(options.check_sorted),
    };

    let by = unwrap_abstract_vec(by.to_rust(cr));

    let agg = unwrap_abstract_vec(agg.to_rust(cr));

    dyn_box!(cr, |lazy_frame, index_column| {
        lazy_frame
            .groupby_dynamic(index_column, by, options)
            .agg(agg)
    })
}

#[ocaml_interop_export]
fn rust_lazy_frame_join(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    other: OCamlRef<DynBox<LazyFrame>>,
    left_on: OCamlRef<OCamlList<DynBox<Expr>>>,
    right_on: OCamlRef<OCamlList<DynBox<Expr>>>,
    how: OCamlRef<JoinType>,
) -> OCaml<DynBox<LazyFrame>> {
    let left_on = unwrap_abstract_vec(left_on.to_rust(cr));
    let right_on = unwrap_abstract_vec(right_on.to_rust(cr));
    let PolarsJoinType(how) = how.to_rust(cr);

    dyn_box!(cr, |lazy_frame, other| {
        lazy_frame.join(other, &left_on, &right_on, JoinArgs::new(how))
    })
}

#[ocaml_interop_export]
fn rust_lazy_frame_sort(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    by_column: OCamlRef<String>,
    descending: OCamlRef<Option<bool>>,
    nulls_last: OCamlRef<Option<bool>>,
    multithreaded: OCamlRef<Option<bool>>,
    maintain_order: OCamlRef<Option<bool>>,
) -> OCaml<DynBox<LazyFrame>> {
    let by_column: String = by_column.to_rust(cr);
    let descending: Option<bool> = descending.to_rust(cr);
    let nulls_last: Option<bool> = nulls_last.to_rust(cr);
    let multithreaded: Option<bool> = multithreaded.to_rust(cr);
    let maintain_order: Option<bool> = maintain_order.to_rust(cr);
    let sort_options: SortOptions = Default::default();
    let sort_options = SortOptions {
        descending: descending.unwrap_or(sort_options.descending),
        nulls_last: nulls_last.unwrap_or(sort_options.nulls_last),
        multithreaded: multithreaded.unwrap_or(sort_options.multithreaded),
        maintain_order: maintain_order.unwrap_or(sort_options.maintain_order),
    };

    dyn_box!(cr, |lazy_frame| lazy_frame.sort(&by_column, sort_options))
}

#[ocaml_interop_export]
fn rust_lazy_frame_vertical_concat(
    cr: &mut &mut OCamlRuntime,
    lazy_frames: OCamlRef<OCamlList<DynBox<LazyFrame>>>,
    rechunk: OCamlRef<bool>,
    parallel: OCamlRef<bool>,
    to_supertypes: OCamlRef<bool>,
) -> OCaml<Result<DynBox<LazyFrame>, String>> {
    let lazy_frames = unwrap_abstract_vec(lazy_frames.to_rust(cr));
    let rechunk = rechunk.to_rust(cr);
    let parallel = parallel.to_rust(cr);
    let to_supertypes = to_supertypes.to_rust(cr);
    concat(
        &lazy_frames,
        UnionArgs {
            rechunk,
            parallel,
            to_supertypes,
        },
    )
    .map(Abstract)
    .map_err(|err| err.to_string())
    .to_ocaml(cr)
}

#[ocaml_interop_export]
fn rust_lazy_frame_diagonal_concat(
    cr: &mut &mut OCamlRuntime,
    lazy_frames: OCamlRef<OCamlList<DynBox<LazyFrame>>>,
    rechunk: OCamlRef<bool>,
    parallel: OCamlRef<bool>,
) -> OCaml<Result<DynBox<LazyFrame>, String>> {
    let lazy_frames = unwrap_abstract_vec(lazy_frames.to_rust(cr));
    let rechunk = rechunk.to_rust(cr);
    let parallel = parallel.to_rust(cr);
    diag_concat_lf(&lazy_frames, rechunk, parallel)
        .map(Abstract)
        .map_err(|err| err.to_string())
        .to_ocaml(cr)
}

#[ocaml_interop_export]
fn rust_lazy_frame_melt(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    id_vars: OCamlRef<OCamlList<String>>,
    value_vars: OCamlRef<OCamlList<String>>,
    variable_name: OCamlRef<Option<String>>,
    value_name: OCamlRef<Option<String>>,
    streamable: OCamlRef<bool>,
) -> OCaml<DynBox<LazyFrame>> {
    let id_vars: Vec<SmartString<LazyCompact>> = id_vars
        .to_rust::<Vec<String>>(cr)
        .into_iter()
        .map(|s| s.into())
        .collect();
    let value_vars: Vec<SmartString<LazyCompact>> = value_vars
        .to_rust::<Vec<String>>(cr)
        .into_iter()
        .map(|s| s.into())
        .collect();
    let variable_name: Option<SmartString<LazyCompact>> = variable_name
        .to_rust::<Option<String>>(cr)
        .map(|s| s.into());
    let value_name: Option<SmartString<LazyCompact>> =
        value_name.to_rust::<Option<String>>(cr).map(|s| s.into());
    let streamable: bool = streamable.to_rust(cr);

    let melt_args = MeltArgs {
        id_vars,
        value_vars,
        variable_name,
        value_name,
        streamable,
    };

    dyn_box!(cr, |lazy_frame| lazy_frame.melt(melt_args))
}

#[ocaml_interop_export(raise_on_err)]
fn rust_lazy_frame_limit(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    n: OCamlRef<OCamlInt>,
) -> OCaml<DynBox<LazyFrame>> {
    let n = n.to_rust::<Coerce<_, i64, u32>>(cr).get()?;

    dyn_box!(cr, |lazy_frame| lazy_frame.limit(n))
}

#[ocaml_interop_export]
fn rust_lazy_frame_explode(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    columns: OCamlRef<OCamlList<DynBox<Expr>>>,
) -> OCaml<DynBox<LazyFrame>> {
    let columns = unwrap_abstract_vec(columns.to_rust(cr));

    dyn_box!(cr, |lazy_frame| lazy_frame.explode(&columns))
}

#[ocaml_interop_export]
fn rust_lazy_frame_with_streaming(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
    toggle: OCamlRef<bool>,
) -> OCaml<DynBox<LazyFrame>> {
    let toggle = toggle.to_rust(cr);

    dyn_box!(cr, |lazy_frame| lazy_frame.with_streaming(toggle))
}

#[ocaml_interop_export]
fn rust_lazy_frame_schema(
    cr: &mut &mut OCamlRuntime,
    lazy_frame: OCamlRef<DynBox<LazyFrame>>,
) -> OCaml<Result<DynBox<Schema>, String>> {
    dyn_box_result!(cr, |lazy_frame| {
        lazy_frame.schema().map(|schema| (*schema).clone())
    })
}
